### PR TITLE
[enhancement](memory) MemCounter supports lock-free thread safety

### PR DESCRIPTION
**Pull request by luozenglin**
# Proposed changes

Issue Number: close #xxx

## Problem summary

make try_add() and update_peak() thread-safe.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by luozenglin** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

File: be/src/runtime/memory/mem_tracker.h [hunk 1]
```diff
@@ -64,32 +64,39 @@ class MemTracker {
...
+            update_peak(v);
         }
 
-        void update_peak() {
-            if (_current_value.load(std::memory_order_relaxed) >
-                _peak_value.load(std::memory_order_relaxed)) {
-                _peak_value.store(_current_value.load(std::memory_order_relaxed),
-                                  std::memory_order_relaxed);
+        void update_peak(int64_t value) {
+            auto pre_value = _peak_value.load(std::memory_order_relaxed);
```
Comment (yiguolei) [inline be/src/runtime/memory/mem_tracker.h:97]: 这个在多线程并发的时候，感觉也不对啊，比如
_peak_value = 3,
线程1： value = 10
线程2： value = 20
两个线程同时运行到line 97， 此时他们拿到的都是pre_value = 3， 此时加入线程1 先执行，那么_peak_value 被更新为10； 现成2 执行 line 98 会直接cas 失败，然后返回

Comment (luozenglin) [reply be/src/runtime/memory/mem_tracker.h:97]: `compare_exchange_weak` will update pre_value on failure and thread 2 will update pre_value to 10 and enter the loop again

**Comment by luozenglin** [comment]
run buildall

File: be/src/runtime/memory/mem_tracker.h (hunk 1 above)
Comment (yiguolei) [reply be/src/runtime/memory/mem_tracker.h:97]: if thread 2 enter the loop again, but the pre_value is still 3, so it will update failed...

Comment (luozenglin) [reply be/src/runtime/memory/mem_tracker.h:97]: compare_exchange_weak updates pre_value if the comparison fails, so next enters the loop the pre_value will be 10

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**